src/pytest_codspeed/plugin.py

```diff
@@ -170,88 +172,58 @@ def pytest_collection_modifyitems(
-    ihook.pytest_runtest_logstart(nodeid=item.nodeid, location=item.location)
-    setup_call = pytest.CallInfo.from_call(
-        lambda: ihook.pytest_runtest_setup(item=item, nextitem=nextitem), "setup"
-    )
-    setup_report = ihook.pytest_runtest_makereport(item=item, call=setup_call)
-    ihook.pytest_runtest_logreport(report=setup_report)
-    reports.append(setup_report)
-    # Run phase
-    if setup_report.passed and not item.config.getoption("setuponly"):
-        assert plugin.lib is not None
-        runtest_call = pytest.CallInfo.from_call(
-            lambda: _run_with_instrumentation(
-                plugin.lib, item.nodeid, item.config, item.runtest
-            ),
-            "call",
-        )
-        runtest_report = ihook.pytest_runtest_makereport(item=item, call=runtest_call)
-        ihook.pytest_runtest_logreport(report=runtest_report)
-        reports.append(runtest_report)
-
-    # Teardown phase
-    teardown_call = pytest.CallInfo.from_call(
-        lambda: ihook.pytest_runtest_teardown(item=item, nextitem=nextitem), "teardown"
-    )
-    teardown_report = ihook.pytest_runtest_makereport(item=item, call=teardown_call)
-    ihook.pytest_runtest_logreport(report=teardown_report)
-    reports.append(teardown_report)
-    ihook.pytest_runtest_logfinish(nodeid=item.nodeid, location=item.location)
-
-    return reports  # Deny further protocol hooks execution
+    testfunction: Callable[P, T],
+) -> Callable[P, T]:
+    @functools.wraps(testfunction)
+    def wrapper(*args: P.args, **kwargs: P.kwargs) -> T:
+        def __codspeed_root_frame__():
+            return testfunction(*args, **kwargs)
+
+        is_gc_enabled = gc.isenabled()
+        if is_gc_enabled:
+            gc.collect()
+            gc.disable()
+        try:
+            if SUPPORTS_PERF_TRAMPOLINE:
+                # Warmup CPython performance map cache
+                __codspeed_root_frame__()
+
+            lib.zero_stats()
+            lib.start_instrumentation()
+            try:
+                return __codspeed_root_frame__()
+            finally:
+                lib.stop_instrumentation()
+                uri = get_git_relative_uri(nodeid, config.rootpath)
+                lib.dump_stats_at(uri.encode("ascii"))
+        finally:
+            if is_gc_enabled:
+                gc.enable()
+
+    return wrapper
+
+
+@pytest.hookimpl(hookwrapper=True)
+def pytest_pyfunc_call(pyfuncitem: pytest.Function) -> Iterator[None]:
+    plugin = get_plugin(pyfuncitem.config)
+    if (
+        plugin.is_codspeed_enabled
+        and should_benchmark_item(pyfuncitem)
+        and not has_benchmark_fixture(pyfuncitem)
+    ):
+        plugin.benchmark_count += 1
+        if plugin.lib is not None and plugin.should_measure:
+            pyfuncitem.obj = wrap_pyfunc_with_instrumentation(
+                plugin.lib,
+                pyfuncitem.nodeid,
+                pyfuncitem.config,
+                pyfuncitem.obj,
+            )
+    yield
```

Comment:
Love this refactor!
However, when a benchmark is defined from a fixture, we should still perform the warmup at this level and not in the benchmark fixture(see the other comment I left on it).